File: backend/pagination.py
"""Small, shared pagination primitives for normalized collection endpoints."""

from math import ceil


DEFAULT_PAGE = 1
DEFAULT_PAGE_SIZE = 20
COMMON_MAX_PAGE_SIZE = 100


def parse_pagination(page=DEFAULT_PAGE, page_size=DEFAULT_PAGE_SIZE, *, max_page_size=COMMON_MAX_PAGE_SIZE):
    """Validate one-based page parameters without silently clamping them."""
    try:
        page = int(page)
        page_size = int(page_size)
    except (TypeError, ValueError) as error:
        raise ValueError("分頁參數必須是整數") from error
    if page < 1:
        raise ValueError("頁碼必須從 1 開始")
    if page_size < 1 or page_size > max_page_size:
        raise ValueError(f"每頁筆數必須介於 1 至 {max_page_size}")
    return page, page_size


def page_result(items, *, total, page, page_size):
    """Build the one canonical collection envelope used by platform pages."""
    total = max(0, int(total))
    total_pages = ceil(total / page_size) if total else 0
    return {
        "items": items,
        "page": page,
        "page_size": page_size,
        "total": total,
        "total_pages": total_pages,
        "has_next": page < total_pages,
        "has_prev": page > 1,
    }

File: frontend/services/pagination.js
"use strict";

/* Shared view-only pagination primitives.  Domain loaders own their query
 * parameters; this helper only normalizes the envelope and renders controls. */
(() => {
  const DEFAULT_PAGE_SIZE = 20;

  function normalize(payload) {
    if (Array.isArray(payload)) {
      return {
        items: payload,
        page: 1,
        page_size: payload.length || DEFAULT_PAGE_SIZE,
        total: payload.length,
        total_pages: payload.length ? 1 : 0,
        has_next: false,
        has_prev: false,
        legacy: true,
      };
    }
    const value = payload || {};
    const items = Array.isArray(value.items) ? value.items : [];
    const page = Math.max(1, Number(value.page ?? 1) || 1);
    const pageSize = Math.max(1, Number(value.page_size ?? value.pageSize ?? DEFAULT_PAGE_SIZE) || DEFAULT_PAGE_SIZE);
    const total = Math.max(0, Number(value.total ?? items.length) || 0);
    const totalPages = Math.max(0, Number(value.total_pages ?? value.totalPages ?? (total ? Math.ceil(total / pageSize) : 0)) || 0);
    return {
      items,
      page,
      page_size: pageSize,
      total,
      total_pages: totalPages,
      has_next: value.has_next ?? value.hasNext ?? page < totalPages,
      has_prev: value.has_prev ?? value.hasPrev ?? page > 1,
      legacy: false,
    };
  }

  function render(payload, { label = "作品分頁", hrefForPage = () => "#" } = {}) {
    const meta = normalize(payload);
    if (meta.total_pages <= 1 && !meta.has_prev && !meta.has_next) return "";
    const link = (page, text, disabled, current = false) => disabled
      ? `<span class="platform-pagination-button is-disabled" aria-disabled="true">${text}</span>`
      : `<a class="platform-button platform-button-small platform-pagination-button" href="${hrefForPage(page)}"${current ? ' aria-current="page"' : ""}>${text}</a>`;
    return `<nav class="platform-pagination" aria-label="${label}"><span class="platform-pagination-status" aria-live="polite">第 ${meta.page} / ${meta.total_pages || 0} 頁 · 共 ${meta.total} 項</span><div class="platform-pagination-actions">${link(meta.page - 1, "上一頁", !meta.has_prev)}<span class="platform-pagination-current" aria-current="page">第 ${meta.page} 頁</span>${link(meta.page + 1, "下一頁", !meta.has_next)}</div></nav>`;
  }

  window.StoryLingoPagination = { normalize, render };
})();
